src/Microsoft.AspNet.Mvc.Core/Internal/CopyOnWriteList.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Mvc.Internal
+{
+    public class CopyOnWriteList<T> : IList<T>
+    {
+        private readonly IReadOnlyList<T> _source;
+        private List<T> _copy;
+
+        public CopyOnWriteList(IReadOnlyList<T> source)
+        {
+            _source = source;
+        }
+
+        protected IReadOnlyList<T> Readable
+        {
+            get
+            {
+                return _copy ?? _source;
+            }
+        }
+
+        protected List<T> Writable
+        {
+            get
+            {
+                if (_copy == null)
+                {
+                    _copy = new List<T>(_source);
+                }
+
+                return _copy;
+            }
+        }
+
+        public T this[int index]
+        {
+            get
+            {
+                return Readable[index];
+            }
+            set
+            {
+                Writable[index] = value;
+            }
+        }
+
+        public int Count
+        {
+            get
+            {
+                return Readable.Count;
+            }
+        }
+
+        public bool IsReadOnly
+        {
+            get
+            {
+                return false;
+            }
+        }
+
+        public void Add(T item)
+        {
+            Writable.Add(item);
+        }
+
+        public void Clear()
+        {
+            Writable.Clear();
+        }
+
+        public bool Contains(T item)
+        {
+            foreach (var obj in Readable)
+            {
+                if (object.Equals(obj, item))
+                {
+                    return true;
+                }
+            }
+
+            return false;
```

Comment:
Why not `Readable.Contains()`? Similar questions for `CopyTo()` and `IndexOf()`.